# 웹 크롤링1 - Static Crawling
#### e-비즈니스학과 201921507 허진아

---

# 1. urllib
- 파이썬은 웹 사이트에 있는 데이터를 추출하기 위해 urllib 라이브러리 사용
- 이를 이용해 HTTP 또는 FTP를 사용해 데이터 다운로드 가능
- urllib은 URL을 다루는 모듈을 모아놓은 패키지
- urllib.request 모듈은 웹 사이트에 있는 데이터에 접근하는 기능 제공, 또는 인증 리다렉트, 쿠키ㅓ럼 인터넷을 이용한 다양한 요청과 처리가 가능

In [5]:
from urllib import request #urllib로 부터 request 패키지를 불러오기

## 1.1 urllib.request를 이용한 다운로드
- urllib.request 모듈에 있는 urlretrive() 함수 이용
- 다음의 코드는 PNG 파일을 test.png라는 이름의 파일로 저장하는 예제임

In [4]:
# 라이브러리 읽어들이기
from urllib import request

url="http://uta.pw/shodou/img/28/214.png"
# url을 불러오기
savename="test.png"  # 'test'라는 이름의 png 확장자 파일을 저장하기

request.urlretrieve(url, savename)  # request 실행하기
print("저장되었습니다")  # 프린트하기

저장되었습니다


## 1.2 urlopen으로 파일에 저장하는 방법
- request.urlopen()은 메모리에 데이터를 올린 후 파일에 저장하게 된다.

In [ ]:
# URL과 저장경로 지정하기
url = "http://uta.pw/shodou/img/28/214.png"
savename = "test1.png"
#다운로드
mem = request.urlopen(url).read()
#파일로 저장하기, wb는 쓰기와 바이너리모드
with open(savename, mode="wb") as f:  # 위에서 정의한 행위를 f라고 하기
    f.write(mem)  
    print("저장되었습니다..")

## 1.3 API 사용하기
### 클라이언트 접속 정보 출력 (기본)
- API는 사용자의 요청에 따라 정보를 반환하는 프로그램
- IP 주소, UserAgent 등 클라이언트 접속정보 출력하는 "IP 확인 API" 접근해서 정보를 추출하는 프로그램

In [ ]:
#데이터 읽어들이기
url="http://api.aoikujira.com/ip/ini"
res=request.urlopen(url)
data=res.read()

#바이너리를 문자열로 변환하기
text=data.decode("utf-8")
print(text)

# 2. BeautifulSoup
- 스크레이핑(Scraping or Crawling)이란 웹사이트에서 데이터를 추출하고, 원하는 정보를 추출하는 것을 의미
- BeautifulSoup이란 파이썬으로 스크레이핑할 대 사용되는 라이브러리로서 HTML/XML 에서 정보를 추출할 수 있도록 도와줌. 그러나 다운로드 기능은 없음,
- 파이썬 라이브러리는 pip 명령어를 이용해 설치 가능. Python Pakage Index(PyPI)에 있는 패키지 명령어를 한줄로 설치 가능
  - URL (http://pypi.python.org/pypi)

### 패키지 import 및 예제 HTML

In [10]:
from bs4 import BeautifulSoup  # BeautifulSoup 모듈 불러오기

In [8]:
html = """
<html><body>
  <h1>스크레이핑이란?</h1>
  <p>웹 페이지를 분석하는 것</p>
  <p>원하는 부분을 추출하는 것</p>
</body></html>
"""

## 2.1 기본 사용
- 다음은 Beautifulsoup를 이용하여 웹사이트로부터 HTML을 가져와 문자열로 만들어 이용하는 예제임
- h1 태그를 접근하기 위해 html-body-h1 구조를 사용하여 soup.html.body.h1 이런식으로 이용하게 됨.
- p 태그는 두개가 있어 soup.html.body.p 한 후 next_sibling을 두번 이용하여 다음 p를 추출. 한번만 하면 그 다음 공백이 추출됨.
- HTML 태그가 복잡한 경우 이런 방식으로 계속 진행하기는 적합하지 않음.

### 2) HTML 분석하기

In [12]:
soup = BeautifulSoup(html, 'html.parser') #html.parser 불러오기

### 3) 원하는 부분 추출하기

In [15]:
h1 = soup.html.body.h1  # body는 h1을 h1로 저장하기
p1 = soup.html.body.p  # body는 p를 p1으로 저장하기
p2 = p1.next_sibling.next_sibling

### 4) 요소의 글자 출력하기

In [16]:
print(f"h1 = {h1.string}")  #h1의 string을 출력해라
print(f"p  = {p1.string}")  #p1의 string을 출력해라
print(f"p  = {p2.string}")  #p2의 string을 출력해라

h1 = 스크레이핑이란?
p  = 웹 페이지를 분석하는 것
p  = 원하는 부분을 추출하는 것


## 2.2 요소를 찾는 method
### 단일 element 추출 : find()
BeautifulSoup는 루트부터 하나하나 요소를 찾는 방법 말고도 find()라는 메소드를 제공함

In [17]:
soup = BeautifulSoup(html, 'html.parser')  # soup라는 변수에 html.parser를 정의해주기

- 1) find()메서드로 원하는 부분 추출하기

In [18]:
title = soup.find("h1")  # h1의 부분을 추출하기
body  = soup.find("p")  # p의 부분을 추출하기
print(title)  # 추출한 h1 부분을 프린트해라

<h1>스크레이핑이란?</h1>


- 2) 텍스트 부분 출력하기

In [19]:
print(f"#title = {title.string}")  # title의 string을 출력해라
print(f"#body = {body.string}")  # body의 string을 출력해라

#title = 스크레이핑이란?
#body = 웹 페이지를 분석하는 것


### 복수 elements 추출 : find_all()
여러개의 태그를 한번에 추출하고자 할때 사용함. 다음의 예제에서는 여러개의 태그를 추출하는 법을 보여주고 있음

In [20]:
html = """
<html><body>
  <ul>
    <li><a href="http://www.naver.com">naver</a></li>
    <li><a href="http://www.daum.net">daum</a></li>
  </ul>
</body></html>
"""

soup = BeautifulSoup(html, 'html.parser')

- 1) find_all() 메서드로 추출하기 

In [21]:
links = soup.find_all("a")  # a의 모든 부분을 추출하기 위해 find를 이용하기
print(links, len(links))

[<a href="http://www.naver.com">naver</a>, <a href="http://www.daum.net">daum</a>] 2


- 2) 링크 목록 출력하기

In [22]:
for a in links:
    href = a.attrs['href'] # href의 속성에 있는 속성값을 추출
    text = a.string 
    print(text, ">", href)

naver > http://www.naver.com
daum > http://www.daum.net


# 3. Css Selector

> Css Selector란, 웹상의 요소에 css를 적용하기 위한 문법으로, 즉 요소를 선택하기 위한 패턴입니다.

출처: https://www.w3schools.com/cssref/css_selectors.asp

앞서 간단하게 태그를 사용하여 데이터를 추출하는 방법에 대해서 살펴보았습니다.

하지만 복잡하게 구조화된 웹 사이트에서 자신이 원하는 데이터를 가져오기 위해서는 Css Selector에 대한 이해가 필요합니다.

서식 | 설명
---- | ---
*    | 모든 요소를 선택
<요소 이름>   | 요소 이름을 기반으로 선택
<클래스 이름>  | 클래스 이름을 기반으로 서낵
#<id 이름>  | id 속성을 기반으로 선택


## BeautifulSoup에서 Css Selector 사용하기
BeautifulSoup에서는 Css Selector로 값을 가져올 수 있도록 find와는 다른 다음과 같은 메서드를 제공합니다.

메서드 | 설명
---- | ---
soup.select_one(선택자)    | CSS 선택자로 요소 하나를 추출합니다.
soup.select(선택자)	   | CSS 선택자로 요소 여러개를 리스트를 추출합니다.


In [25]:
html = """
<html><body>
<div id="meigen">
  <h1>위키북스 도서</h1>
  <ul class="items">
    <li>유니티 게임 이펙트 입문</li>
    <li>스위프트로 시작하는 아이폰 앱 개발 교과서</li>
    <li>모던 웹사이트 디자인의 정석</li>
  </ul>
</div>
</body></html>
"""

# HTML 분석하기 
soup = BeautifulSoup(html, 'html.parser')

- 필요한 부분을 CSS 쿼리로 추출하기

In [26]:
# 타이틀 부분 추출하기 --- (※3)
h1 = soup.select_one("div#meigen > h1").string
print(f"h1 = {h1}")

# 목록 부분 추출하기 --- (※4)
li_list = soup.select("div#meigen > ul.items > li")
for li in li_list:
  print(f"li = {li.string}")

h1 = 위키북스 도서
li = 유니티 게임 이펙트 입문
li = 스위프트로 시작하는 아이폰 앱 개발 교과서
li = 모던 웹사이트 디자인의 정석


# 4. 활용 예제
앞서 배운 urllib과 BeautifulSoup를 조합하면, 웹스크레이핑 및 API 요청 작업을 쉽게 수행하실 수 있습니다.

1. URL을 이용하여 웹으로부터 html을 읽어들임 (urllib)
2. html 분석 및 원하는 데이터를 추출 (BeautifulSoup)

In [27]:
from bs4 import BeautifulSoup  # bs4로부터 beautifulSoup 불러오기
from urllib import request, parse  # urllib로 부터 request와 parse 불러오기

## 4.1 네이버 금융 - 환율 정보
- 다양한 금융 정보가 공개돼 있는 "네이버 금융"에서 원/달러 환율 정보를 추출해보자!
- 네이버 금융의 시장 지표 페이지 https://finance.naver.com/marketindex/
- 다음은 원/달러 환율 정보를 추출하는 프로그램임

### 1) HTML 가져오기

In [31]:
url = "https://finance.naver.com/marketindex/"  # 스크레이핑 하기 위한 url을 정의해주기
res = request.urlopen(url)  # 위에서 정의해준 url을 request 해주기

### 2) HTML 분석하기

In [34]:
soup = BeautifulSoup(res, "html.parser")  # soup라는 변수에 html.parser를 정의해주기

### 3) 원하는 데이터 추출하기

In [35]:
price = soup.select_one("div.head_info > span.value").string
# 원하는 데이터를 추출하기 위해 Css Selector를 div.head_info 그리고 span.value 순ㅅ서대로 접근하여 string을 선택하기
print("usd/krw =", price) # 환율 정보임을 알려주기 위해 price가 의미하는 문자열을 앞에 함께 출력해주기

usd/krw = 1,184.00


## 4.2 기상청 RSS
- 기상청 RSS에서 특정 내용을 추출하는 예제
- 기상청 RSS에서 XML 데이터를 추출하고 XML 내용을 출력
- 기상청의 RSS 서비스에 지역 번호를 지정하여 데이터 요청해보기  
http://www.kma.go.kr/weather/forecast/mid-term-rss3.jsp
  - 참고: 기상청 RSS http://www.kma.go.kr/weather/lifenindustry/service_rss.jsp

매개변수 | 의미
---- | ---
stnid   | 기상정보를 알고 싶은 지역을 지정


- 지역번호는 다음과 같음

지역 | 지역번호 | 지역 | 지역번호
---- | --- | --- | ---
전국 | 108 | 전라북도 | 146
서울/경기도 | 109 | 전라남도 | 156
강원도 | 105 | 경상북도 | 143
충청북도 | 131 | 경상남도 | 159
충청남도 | 133 | 제주특별자치도 | 184

- 파이썬으로 요청 전용 매개변수를 만들 때는 urllib.parse 모듈의 urlencode() 함수를 사용해 매개변수를 URL로 인코딩한다.

### 1) HTML 가져오기

In [43]:
url = "http://www.kma.go.kr/weather/forecast/mid-term-rss3.jsp"

#매개변수를 URL로 인코딩한다.
values = {
    'stnId':'109'
}

params=parse.urlencode(values)
url += "?"+params # URL에 매개변수 추가
print("url=", url)

res = request.urlopen(url)

url= http://www.kma.go.kr/weather/forecast/mid-term-rss3.jsp?stnId=109


### 2) HTML 분석하기

In [44]:
soup = BeautifulSoup(res, "html.parser")  # soup라는 변수에 html.parser를 정의해주기

### 3) 원하는 데이터 추출하기

In [45]:
header = soup.find("header")  # header을 찾아 header에 정의해주기

title = header.find("title").text  # title을 찾아 변수 title에 정의해주기
wf = header.find("wf").text  # wf를 찾아 wf에 정의해주기

print(title)
print(wf)

서울,경기도 육상중기예보
○ (기온) 이번 예보기간 아침최저기온은 13~19도, 낮최고기온은 24~27도로 어제(20일, 아침최저기온 15~20도, 낮최고기온 27~29도)보다 낮겠습니다.<br />○ (해상) 서해중부해상의 물결은 0.5~2.0m로 일겠습니다.<br />○ (주말전망) 25일(토)과 26일(일)은 구름많겠습니다. 아침최저기온은 14~18도, 낮최고기온은 25~27도가 되겠습니다.


- css selector 기반

In [46]:
title = soup.select_one("header > title").text  # Css Selector의 순서대로 찾고 text로 추출하기
wf = header.select_one("header wf").text  # header에서 wf를 찾고 text로 추출하기

print(title)
print(wf)

서울,경기도 육상중기예보
○ (기온) 이번 예보기간 아침최저기온은 13~19도, 낮최고기온은 24~27도로 어제(20일, 아침최저기온 15~20도, 낮최고기온 27~29도)보다 낮겠습니다.<br />○ (해상) 서해중부해상의 물결은 0.5~2.0m로 일겠습니다.<br />○ (주말전망) 25일(토)과 26일(일)은 구름많겠습니다. 아침최저기온은 14~18도, 낮최고기온은 25~27도가 되겠습니다.


## 4.3 윤동주 작가의 작품 목록
- 위키문헌 (https://ko.wikisource.org/wiki) 에 공개되어 있는 윤동주의 작품목록을 가져오기
- 윤동주 위키  (https://ko.wikisource.org/wiki/%EC%A0%80%EC%9E%90:%EC%9C%A4%EB%8F%99%EC%A3%BC)
- 하늘과 바람과 시 부분을 선택한 후 오른쪽 마우스 이용해 copy selector로 카피하면 다음의 CSS 선택자가 카피됨
  - #mw-content-text > div > ul:nth-child(6) > li > b > a
- nth-child(n) 은 n 번째 요소를 의미 즉 6번째 요소를 의미, #mw-content-text 내부에 있는 url 태그는 모두 작품과 관련된 태그. 따라서 따로 구분할 필요는 없으며 생략해도 됨. BeautifulSoup는 nth-child 지원하지 않음
  - Recall PR7 Problem1

In [47]:
# 뒤의 인코딩 부분은 "저자:윤동주"라는 의미입니다.
# 따로 입력하지 말고 위키 문헌 홈페이지에 들어간 뒤에 주소를 복사해서 사용하세요.

url = "https://ko.wikisource.org/wiki/%EC%A0%80%EC%9E%90:%EC%9C%A4%EB%8F%99%EC%A3%BC"
res = request.urlopen(url)
soup = BeautifulSoup(res, "html.parser")

# #mw-content-text 바로 아래에 있는 
# ul 태그 바로 아래에 있는
# li 태그 아래에 있는
# a 태그를 모두 선택합니다.
a_list = soup.select("#mw-content-text   ul > li  a")
for a in a_list:
    name = a.string
    print(f"- {name}", )

- 하늘과 바람과 별과 시
- 증보판
- 서시
- 자화상
- 소년
- 눈 오는 지도
- 돌아와 보는 밤
- 병원
- 새로운 길
- 간판 없는 거리
- 태초의 아침
- 또 태초의 아침
- 새벽이 올 때까지
- 무서운 시간
- 십자가
- 바람이 불어
- 슬픈 족속
- 눈감고 간다
- 또 다른 고향
- 길
- 별 헤는 밤
- 흰 그림자
- 사랑스런 추억
- 흐르는 거리
- 쉽게 씌어진 시
- 봄
- 참회록
- 간(肝)
- 위로
- 팔복
- 못자는밤
- 달같이
- 고추밭
- 아우의 인상화
- 사랑의 전당
- 이적
- 비오는 밤
- 산골물
- 유언
- 창
- 바다
- 비로봉
- 산협의 오후
- 명상
- 소낙비
- 한난계
- 풍경
- 달밤
- 장
- 밤
- 황혼이 바다가 되어
- 아침
- 빨래
- 꿈은 깨어지고
- 산림
- 이런날
- 산상
- 양지쪽
- 닭
- 가슴 1
- 가슴 2
- 비둘기
- 황혼
- 남쪽 하늘
- 창공
- 거리에서
- 삶과 죽음
- 초한대
- 산울림
- 해바라기 얼굴
- 귀뚜라미와 나와
- 애기의 새벽
- 햇빛·바람
- 반디불
- 둘 다
- 거짓부리
- 눈
- 참새
- 버선본
- 편지
- 봄
- 무얼 먹구 사나
- 굴뚝
- 햇비
- 빗자루
- 기왓장 내외
- 오줌싸개 지도
- 병아리
- 조개껍질
- 겨울
- 트루게네프의 언덕
- 달을 쏘다
- 별똥 떨어진 데
- 화원에 꽃이 핀다
- 종시


# 일반 문제
---

In [48]:
from bs4 import BeautifulSoup
from urllib import request

## 1. 네이터 뉴스 헤드라인
배운 내용을 바탕으로 네이버 뉴스(https://news.naver.com/)에서 헤드라인 뉴스의 제목을 추출해보고자 합니다.
> ***Q*** : 다음의 코드에 css selector를 추가하여 최신 기사의 헤드라인을 스크레이핑하는 코드를 완성하시오.

In [49]:
url = "https://news.naver.com/"  # 원하는 정보를 스크레이핑 하기 위한 url을 정의하기

res = request.urlopen(url)  # 변수 res에 정의한 url을 다시 열고 request 하기
soup = BeautifulSoup(res, "html.parser")  # soup라는 변수에 html.parser를 정의해주기

selector = "#today_main_news > div.hdline_news > ul > li > div.hdline_article_tit > a"
# 추출하고자 하는 Css Selector를 정의해주기

for a in soup.select(selector):
    title = a.text
    print(title)


                                        김정숙 여사, BTS와 뉴욕 메트로폴리탄 미술관 방문…"K컬처가 외교사절 역할"
                                    

                                        카카오 김범수, ‘최고 부자’ 자리 이재용에게 다시 내줘
                                    

                                        코로나19 백신 접종 안 하면 미국 못 간다
                                    

                                        “장제원 아들 노엘, 힙합계 명예 실추…퇴출해야” 성명
                                    

                                        코로나19 신규확진 1729명, 나흘째 ‘요일 최다’
                                    


# 2. 시민의 소리 게시판
다음은 서울시 대공원의 시민의 소리 게시판 입니다.  

https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgList.do?pgno=1

해당 페이지에 나타난 게시글들의 제목을 수집하고자 합니다.

> ***Q***: 다음의 코드에 css selector를 추가하여 해당 페이지에서 게시글의 제목을 스크레이핑하는 코드를 완성하시오. 또한 과제 제출시 하단의 추가 내용을 참고하여 수집한 데이터를 csv 형태로 저장하여 해당 csv 파일도 함께 제출하시오.

In [50]:
url_head = "https://www.sisul.or.kr"  # 시민의 소리 게시물의 제목을 스크레이핑 하기 위한 url을 입력하기

url_board = url_head + "/open_content/childrenpark/qna/qnaMsgList.do?pgno=1"
# 시민의 소리 게시물의 제목 다음으로 추출해야 하는 내용의 Css Selector를 연결하기


res = request.urlopen(url_board)
soup = BeautifulSoup(res, "html.parser")

# selector = "#detail_con > div.generalboard > table > tbody > tr > td.left.title > a"
selector = "#detail_con > div.generalboard > table > tbody > tr > td.left.title > a"
titles = []
links = []
for a in soup.select(selector):  # 순서대로 접근하기
    titles.append(a.text)  # a의 text를 titles에 더해주기
    links.append(url_head + a.attrs["href"])  # url_head와 연결하여 더해주기
    
print(titles, links)

    

['강창수 해설사님 ', '동물해설사님 칭찬', '강창수 동물 해설사님', '놀이동산 푸드코트 김치가 중국산인 이유는?', '주슨트 설명 최고예요!!', '강창수 주슨트님 최고 !!', 'ZOOCENT 스케줄표?', '호주동물 호주설명 ', '호주및 호주동물 설명에 대해 ', '동물원'] ['https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgDetail.do;jsessionid=TWbuXPFkGypc810fqpcu8E42AaQJObpq1P1ZD6U9LjYC8fDEp11Qo1oStNyZMtHO.etisw1_servlet_user?qnaid=QNAS20210920000001&pgno=1', 'https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgDetail.do;jsessionid=TWbuXPFkGypc810fqpcu8E42AaQJObpq1P1ZD6U9LjYC8fDEp11Qo1oStNyZMtHO.etisw1_servlet_user?qnaid=QNAS20210919000004&pgno=1', 'https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgDetail.do;jsessionid=TWbuXPFkGypc810fqpcu8E42AaQJObpq1P1ZD6U9LjYC8fDEp11Qo1oStNyZMtHO.etisw1_servlet_user?qnaid=QNAS20210919000003&pgno=1', 'https://www.sisul.or.kr/open_content/childrenpark/qna/qnaMsgDetail.do;jsessionid=TWbuXPFkGypc810fqpcu8E42AaQJObpq1P1ZD6U9LjYC8fDEp11Qo1oStNyZMtHO.etisw1_servlet_user?qnaid=QNAS20210918000002&pgno=1', 'https://www.sisul.or.kr/open_con

## 추가내용
수집된 자료를 데이터프레임으로 만들어 csv로 저장하는 것이 일반적입니다.

In [53]:
import pandas as pd  # 모듈 pandas를 불러온 후 pd로 부르기로 하기

board_df = pd.DataFrame({"title": titles, "link": links})
# 결과값을 표(데이터프레임)의 형태로 추출하기 위해 데이터 프레임화 해주기
board_df.head()  # 상위 5개만 추출하기


title                                               link
0                강창수 해설사님   https://www.sisul.or.kr/open_content/childrenp...
1                동물해설사님 칭찬  https://www.sisul.or.kr/open_content/childrenp...
2              강창수 동물 해설사님  https://www.sisul.or.kr/open_content/childrenp...
3  놀이동산 푸드코트 김치가 중국산인 이유는?  https://www.sisul.or.kr/open_content/childrenp...
4            주슨트 설명 최고예요!!  https://www.sisul.or.kr/open_content/childrenp...

In [52]:
board_df.to_csv("board.csv", index=False)